### ISSUE TITLE
Feature Request: Option to get rid of Checkbox intermediate state or change toggle order

### ISSUE THREAD

**Issue by KnowZero** [quasarframework/quasar#2067]
Quasar: 0.15-0.16

It would be nice to have the option of completely removing the intermediate state in checkboxes or maybe set the toggle order of states.

I know I can set the 'data' to the false-value or true-value but that isn't very convenient when working with dynamic forms.

Even if I set something like indeterminate-value="none" and don't specify a false-value, it still picks the indeterminate-value. (Even though it should technically pick the false value since that is null and not the intermediate value)

I would be able to get around this if I just change the indeterminate-icon="check_box_outline_blank", but then the user has to click twice to turn it on since I can't control the toggle order.

So the only way I can achieve close to the needed functionality is to flip everything around:
false-value="yes" indeterminate-icon="check_box_outline_blank" unchecked-icon="check_box" checked-icon="check_box_outline_blank" keep-color

But having false-value as the true value isn't a very clean solution.

https://jsfiddle.net/wso1av4f/5/

**Comment by rstoenescu** [comment]
Hi,

Maybe I'm missing some point, but are you aware of the Boolean `toggle-indeterminate` prop? If you don't want the indeterminate state, just set it to Boolean "false"..

**Comment by KnowZero** [comment]
@rstoenescu  How do I do that? I tried

toggle-indeterminate="false" and :toggle-indeterminate="false" and they don't seem to work. Am I doing something wrong?

**Comment by KnowZero** [comment]
@rstoenescu From my understanding, toggle-indeterminate is false by default. But even if it is false by default, it still exists until user clicks once. But there should be an option to completely eliminate it. Look at my jsfiddle, the last checkbox is what I would like:

https://jsfiddle.net/wso1av4f/5/

**Comment by gaby64** [comment]
I do need a null-is-false prop

**Comment by rstoenescu** [comment]
@gaby64 `:false-value="null" indeterminate-value="something-else"`

**Comment by gaby64** [comment]
I have tried that, they are indeterminate on render

**Comment by rstoenescu** [comment]
Works fine: https://codepen.io/rstoenescu/pen/YzPJYdv?editable=true&editors=101
Correct the initial value of the model... @gaby64

**Comment by gaby64** [comment]
undefined is my correct inital value for an object array in a v-for

**Comment by gaby64** [comment]
https://codepen.io/gaby64/pen/YzPREwm?editable=true&editors=101


the only change required to make it work is 

`props: {
    value: {
      default: null
    }`

in checkbox.js

**Comment by smolinari** [comment]
Undefined isn't a proper value. It isn't a value. You aren't saving undefined in your database, are you?

Scott

**Comment by gaby64** [comment]
> Undefined isn't a proper value. It isn't a value. You aren't saving undefined in your database, are you?
> 
> Scott

no

the point is its a temporary value, only used to mark the items as selected

once the chekbox is clicked, item.selected is a valid observable property.

**Comment by smolinari** [comment]
> the point is its a temporary value, only used to mark the items as selected

But, undefined won't mark the items selected and it makes no sense to do so. (unselected either). If is a case of needing to "preset" the value, you could just fill your data with a proper checked or unchecked value. 

Scott

**Comment by gaby64** [comment]
> > the point is its a temporary value, only used to mark the items as selected
> 
> But, undefined won't mark the items selected and it makes no sense to do so. (unselected either). If is a case of needing to "preset" the value, you could just fill your data with a proper checked or unchecked value.
> 
> Scott

true defines them as selected, undefined is not selected
you may want to waste cycles on such things when the answer is easily what I proposed and as works in other ui frameworks

**Comment by smolinari** [comment]
I digress. We obviously have differing points of view on how to set up this component and its state. 

Scott

**Comment by leosdad** [comment]
Just guessing, but maybe the OP wanted the same as I do when mewntioning a "toggle order". The current toggle cycle is _indeterminate_ --> _unckecked_ --> _checked_. It would be very nice to have an option to invert the cycle, i.e. **_indeterminate_** --> _**checked**_ --> _**unckecked**_.

Motivation: when the checkbox default state is indeterminate, clicking it once would **check** the option instead of unchecking it, which would arguably feel more natural to users.

**Comment by rstoenescu** [comment]
Toggle order available in v1.12

**Comment by leosdad** [comment]
Yes, thank you! A much welcome change.

**Comment by leosdad** [comment]
I need any falsy value (false, null, zero, undefined) to be displayed as unckecked. This is a common requisite when working with JSON data. Let's say I have three checkboxes that are bound to 3 variables, and one of them is omitted:

```
{
  "var1": false,
  // var2 is omitted
  "var3": null,
}
```

In this example I'd like all 3 values to show up as unckecked. Is this currently possible? Thanks.

**Comment by leosdad** [comment]
I've just created a component like this:

```js
Vue.component('MyCheckbox',
{
  template: `
    <q-checkbox
      :value="!!$attrs.value"
      v-bind="$attrs"
      v-on="$listeners"
      />
  `
});
```

This kinda does what I want, but I'm not sure this is the best answer.

**Comment by Flamenco** [comment]
I ran into this issue because I needed a quick way to reset my state.  The problem is in the implementation of `v-model` for the checkbox.  I also did not want to persist a bunch of false values when there were no items selected.

So I solved it by going `old-school` and using `:modelValue`, `@update:modelValue`, and converting a falsey value to a boolean.

const multiSelectedItems = ref({})

function clearSelection() {
 multiSelectedItems.value = {}
}


And in the template
```
<q-checkbox
  :model-value="!!multiSelectedItems[id]"
  @update:model-value="multiSelectedItems[item.id]=!multiSelectedItems[id]"
/>
```
